4단계 - Interceptor 구현

iltaek opened the pull request:
안녕하세요

jdbc 마지막 단계네요.
지난 번 피드백 반영하고 이번 단계 요구사항 구현했습니다.

이번에도 잘 부탁드립니다. :)

iltaek committed:
feat($test): JdbcTemplateTest 추가

iltaek committed:
feat($production): 4단계 - Interceptor 구현
- Interceptor 인터페이스 및 구현체(StopWatchInterceptor) 추가
- 추가한 Interceptor 를 HandlerInterceptorRegistry 로 DispatcherServlet에 적용

boorownie commented on src/main/java/core/mvc/HandlerInterceptorRegistry.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Created by iltaek on 2020/07/07 Blog : http://blog.iltaek.me Github : http://github.com/iltaek
+ */
+public class HandlerInterceptorRegistry {
```
Interceptor를 스택 형식으로 호출이 되도록 구현할 수 있을까요?

iltaek replied:
네 무슨 말하시는지 알 것 같습니다. 수정해서 다시 PR 할게요!

boorownie commented on src/main/java/core/mvc/interceptor/StopWatchInterceptor.java:
```diff
@@ -0,0 +1,28 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(StopWatchInterceptor.class);
+
+    private final ThreadLocal<Long> startedTime = new ThreadLocal<>();
```
스레드로컬 활용 👍

boorownie commented on src/main/java/core/mvc/interceptor/StopWatchInterceptor.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Created by iltaek on 2020/07/07 Blog : http://blog.iltaek.me Github : http://github.com/iltaek
+ */
+public class StopWatchInterceptor implements Interceptor {
```
Interceptor 활용 요구사항 확인 👍

boorownie commented on src/test/java/core/jdbc/JdbcTemplateTest.java:
```diff
@@ -0,0 +1,86 @@
+/**
+ * Created by iltaek on 2020/07/12 Blog : http://blog.iltaek.me Github : http://github.com/iltaek
+ */
+class JdbcTemplateTest {
```
피드백 반영 확인 👍

boorownie reviewed (changes requested):
3단계 피드백 반영 및 4단계 요구사항 구현 확인했습니다 👍 
요구사항에 맞게 잘 구현해주셨네요 :)
소소한 피드백 남겨보았는데요~ 한번 고민해보시면 좋을 것 같습니다!ㅎㅎ

iltaek committed:
feat($production): Interceptor 들이 스택 형식으로 호출되도록 수정

- preInterceptors, postIntercepters 를 두어 preHandle, postHandle 의 처리 순서가 짝이 맞도록 수정. (ex. pre 1->2->3, post 3->2->1)

iltaek committed:
feat($production): InterceptorRegistry 가 관리하는 interceptors 스택에 쓰레드로컬 적용

- Interceptor의 처리 순서가 중요하기 때문에 이를 관리하는 스택을 ThreadSafe하게 ThreadLocal 처리한다.

iltaek commented:
Interceptor의 처리 순서를 맞춰주기위해 스택을 사용했습니다.
스택에 쓰레드로컬을 적용하는 것에 고민을 좀 했었는데요
여러 리퀘스트 처리 쓰레드가 동시에 동작하면 결국 스택에 pop/push 하는 과정이 꼬일 수 있을 것 같아서
쓰레드로컬 적용했습니다.

boorownie commented on src/main/java/core/mvc/HandlerInterceptorRegistry.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class HandlerInterceptorRegistry {
+
+    private static final ThreadLocal<Stack<Interceptor>> preInterceptors = new ThreadLocal<>();
+    private static final ThreadLocal<Stack<Interceptor>> postInterceptors = new ThreadLocal<>();
```
ThreadLocal을 활용한 Interceptor 관리 👍

boorownie reviewed (approved):
피드백 요구사항 모두 잘 반영해주셨습니다 💯 
JDBC미션 수행하시느라 고생 많으셨어요!
남은 미션도 모두 끝까지 잘 수행해주셔서 좋은 경험 얻으시길 기원 할게요!
머지하겠습니다~~
